src/renderer/wg_engine/tvgWgCompositor.cpp

```diff
@@ -0,0 +1,509 @@
+        composeImage(context, renderData, &storageClipPath, CompositeMethod::ClipPath);
+    // use custom blending
+    } else if (blentType == WgPipelineBlendType::Custom)
+        blendImage(context, renderData, blendMethod);
+    else // use direct hardware blending
+        drawImage(context, renderData, blentType);
+};
+
+
+void WgCompositor::blendShape(WgContext& context, WgRenderDataShape* renderData, BlendMethod blendMethod)
+{
+    assert(renderData);
+    assert(commandEncoder);
+    assert(currentTarget);
+    // skip shape blending if they are not exists
+    if (renderData->renderSettingsShape.skip) return;
+    if (renderData->meshGroupShapes.meshes.count == 0) return;
+    // store current render pass
+    WgRenderStorage *target = currentTarget;
+    endRenderPass();
+    // render into intermediate buffer
+    beginRenderPass(commandEncoder, &storageInterm, true);
+    drawShape(context, renderData, WgPipelineBlendType::Custom);
+    endRenderPass();
+    // run blend
+    blend(commandEncoder, &storageInterm, target, 255, blendMethod, renderData->renderSettingsShape.rasterType);
+    // restore current render pass
+    beginRenderPass(commandEncoder, target, false);
+}
+
+void WgCompositor::blendStrokes(WgContext& context, WgRenderDataShape* renderData, BlendMethod blendMethod)
+{
+    assert(renderData);
+    assert(commandEncoder);
+    assert(currentTarget);
+    // skip strokes blending if they are not exists
+    if (renderData->renderSettingsStroke.skip) return;
+    if (renderData->meshGroupStrokes.meshes.count == 0) return;
+    // store current render pass
+    WgRenderStorage *target = currentTarget;
+    endRenderPass();
+    // render into intermediate buffer
+    beginRenderPass(commandEncoder, &storageInterm, true);
+    drawStrokes(context, renderData, WgPipelineBlendType::Custom);
+    endRenderPass();
+    // run blend
+    blend(commandEncoder, &storageInterm, target, 255, blendMethod, renderData->renderSettingsStroke.rasterType);
+    // restore current render pass
+    beginRenderPass(commandEncoder, target, false);
+};
+
+
+void WgCompositor::blendImage(WgContext& context, WgRenderDataPicture* renderData, BlendMethod blendMethod)
+{
+    assert(renderData);
+    assert(commandEncoder);
+    assert(currentTarget);
+    // store current render pass
+    WgRenderStorage *target = currentTarget;
+    endRenderPass();
+    // render into intermediate buffer
+    beginRenderPass(commandEncoder, &storageInterm, true);
+    drawImage(context, renderData, WgPipelineBlendType::Custom);
+    endRenderPass();
+    // run blend
+    blend(commandEncoder, &storageInterm, target, 255, blendMethod, WgRenderRasterType::Image);
+    // restore current render pass
+    beginRenderPass(commandEncoder, target, false);
+};
+
+
+void WgCompositor::composeShape(WgContext& context, WgRenderDataShape* renderData, WgRenderStorage* msk, CompositeMethod composeMethod)
+{
+    assert(msk);
```

Comment:
can we use a name 'mask' instead 'msk' ?

Reply:
its the same name as in the shaders, but yes, we can